Firmware libraries settings overhaul

elad335 opened the pull request:
* Remove all lib loading types. E.g. "liblv2.sprx only" etc, replace their usage with two lists. Empty lists is default fw loading behaviour.
Selecting an SPRX file via HLE-default list will LLE it, selecting an SPRX file via HLE-default list will LLE it.
As a nice side effect makes HLEing of libs simpler than ever by selecting them in list such as SPURS or GCM. (if they are LLEd by default).
Previously you had to use manual list type and randomly select LLE-by-defaults libs until the game works, only then unselect the lib you want to HLE.
* Minor libsysmodule.sprx loading fix.
* Fixed LLVM compilation of firmware. (simply didn't compile after installation)
* Handle invalid settings in EnhanceRadioButton (GUI button groups), select default setting button instead of selecting nothing.
* Fixed g_prx_list sorting by using std::map instead of std::unordered_map.
* Fixed PS1 classics. fixes #9394 
* Added a reset button for libraries list.

MISC:
closes #6442

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -611,17 +612,34 @@ void emu_settings::EnhanceRadioButton(QButtonGroup* button_group, emu_settings_t
 	}
 
-	for (int i = 0; i < options.count(); i++)
+	for (int i = 0, unchecked = 1, def_pos = -1; i < options.count() + unchecked; i++)
```
I'd like to move the default setter out of this loop. I found it rather confusing to read.
Something like:
```
bool found = false;
int def_pos = -1;

// loop () {}

if (!found && def_pos >= 0)
{
	button_group->button(def_pos)->setChecked(true);
}
```

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -611,17 +612,34 @@ void emu_settings::EnhanceRadioButton(QButtonGroup* button_group, emu_settings_t
 			button_group->button(i)->setChecked(true);
 		}
 
+		if (def_pos == -1 && options[i] == def)
```
I think this can safely be an else if

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1057,12 +1053,40 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 
 	ui->searchBox->setPlaceholderText(tr("Search libraries", "Library search box"));
+	ui->searchBox->setEnabled(true);
+	ui->lleList->setEnabled(true);
```
these are redundant

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1057,12 +1053,40 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+			const auto text = sstr(items[i]->text());
+
+			// Select default LLE sprx list for manual lib loading, otherwise all libs are unchecked
+			items[i]->setCheckState(id == static_cast<int>(lib_loading_type::manual) && g_prx_list.count(text) && !g_prx_list.at(text)? Qt::Checked : Qt::Unchecked);
```
This is more readable imo:
```suggestion
			const bool selected = id == static_cast<int>(lib_loading_type::manual) && g_prx_list.count(text) && !g_prx_list.at(text);
			items[i]->setCheckState(selected ? Qt::Checked : Qt::Unchecked);
```

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1057,12 +1053,40 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+
+		for (uint i = 0; i < items.size(); i++)
+		{
+			ui->lleList->addItem(items[i]);
```
There is an undefined behavior:
The search filter is not considered when changing the lib mode...

elad335 replied:
Fixed, fixed, fixed, fixed, fixed and fixed.

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1057,12 +1053,40 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+
+		for (uint i = 0; i < items.size(); i++)
+		{
+			const auto text = sstr(items[i]->text());
```
I think the auto type should be reserved for complicated types.
This only makes it harder to read and review.

```suggestion
			const std::string text = sstr(items[i]->text());
```

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -611,22 +612,37 @@ void emu_settings::EnhanceRadioButton(QButtonGroup* button_group, emu_settings_t
 		});
 	}
+
+	if (!found)
```
Missing check for count > 0.
Maybe check that in the early out above

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1094,6 +1083,44 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+			items.push_back(ui->lleList->item(i)->clone());
+		}
+
+		for (uint i = 0; i < items.size(); i++)
```
I'd simply integrate this into the other lambda and just get the text and button Id in the lambda itself or as arg.

elad335 replied:
Fixed, fixed.

elad335 commented:
Something inflated azure build heap memory usage...

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1107,11 +1105,6 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 
-	if (const int lib_mode_button_id = lib_mode_bg->checkedId(); lib_mode_button_id >= 0)
-	{
-		on_lib_button_clicked(lib_mode_button_id);
```
this needs to be called anyway, since we want to presort the libs

elad335 replied:
1 - We dont need to sort them anymore because they are sorted in g_prx_list.
2 - Calling it invalidates selection state of the checkboxes which we only want after switching mode.

Megamouse replied:
alright

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1087,6 +1076,15 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 
 		for (uint i = 0; i < items.size(); i++)
 		{
+			if (id >= 0)
```
do this in the loop above, otherwise sorting is broken

elad335 replied:
Fixed.

elad335 commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1095,8 +1093,8 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 	// Events
-	connect(lib_mode_bg, &QButtonGroup::idClicked, on_lib_button_clicked);
-	connect(ui->searchBox, &QLineEdit::textChanged, on_search_box_text_changed);
+	connect(lib_mode_bg, &QButtonGroup::idToggled, [this, on_lib_state_changed](int id){ on_lib_state_changed(ui->searchBox->text(), id); });
```
idToggled is more appropriate here I think.

Megamouse replied:
A: the function args are wrong.
B: no it is not. with this you'll call state changed on programmatical changes, which we usually want to prevent

Megamouse replied:
to clarify:
Qt itself calls click() when the user changes selection (for example by mouse or keyboard).
If you now use toggle, you'll trigger the signal whenever we change the value in code, which might add weird errors in the future. We then would have to add a signal blocker when we don't want the slot being called....

So it's much easier to just call the slot ourselves if necessary after we are finished toggling buttons.

elad335 replied:
Fixed lambda args, thanks. I want to invalidate selection state *only* after switching mode and not just random clicks on the same setting so I think idToggled is better for this..

elad335 replied:
oh hmm I'll think of something better without this signal.

Megamouse replied:
if you're worried about selection state during the dialog then you're doing something wrong.
The global library state should only be altered on apply_configs.
Calling the lambda with the current selection after initialization should not have any effect.

elad335 replied:
Fixed.

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -611,22 +612,36 @@ void emu_settings::EnhanceRadioButton(QButtonGroup* button_group, emu_settings_t
 		});
 	}
+
+	if (!found && def_pos >= 0)
+	{
+		// Select the default option on invalid setting string
+		button_group->button(def_pos)->setChecked(true);
+	}
```
I noticed that we are missing something here:

```suggestion
	if (!found)
	{
		ensure(def_pos >= 0);

		cfg_log.error("EnhanceRadioButton'%s' tried to set an invalid value: %s. Setting to default: %s.", cfg_adapter::get_setting_name(type), sstr(selected), sstr(def));
		m_broken_types.insert(type);

		// Select the default option on invalid setting string
		button_group->button(def_pos)->setChecked(true);
	}
```

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -611,22 +612,41 @@ void emu_settings::EnhanceRadioButton(QButtonGroup* button_group, emu_settings_t
+	{
+		ensure(def_pos >= 0);
+
+		cfg_log.error("EnhanceRadioButton'%s' tried to set an invalid value: %s. Setting to default: %s.", cfg_adapter::get_setting_name(type), sstr(selected), sstr(def));
```
sorry, I made a typo during copy pasta
```suggestion
		cfg_log.error("EnhanceRadioButton '%s' tried to set an invalid value: %s. Setting to default: %s.", cfg_adapter::get_setting_name(type), sstr(selected), sstr(def));
```

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.h:
```diff
@@ -60,6 +60,9 @@ private Q_SLOTS:
 	QString m_discord_state;
 
+	// Lib loading state
+	int last_lib_mode = -1;
```
Ah, now I understand. You meant the reset of unapplied values when clicking the same button after changing the libs.
Anyway, here some advise:
```suggestion
	int m_last_lib_mode = -1;
```

elad335 replied:
Fixed.

Megamouse reviewed (approved):
I guess it's fine now from a pure GUI side of things

elad335 commented:
* liblv2 and libsysmodule modes now hide all irrelavant "LLE-by-default" sprx files. (it didnt matter if their checkbox was selected or not on master, LLE was always used)

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1085,18 +1087,42 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+			const std::string text = sstr(qtext);
+
+			const bool lle_default = g_prx_list.count(text) && !g_prx_list.at(text);
+			const bool hidden = !qtext.contains(search_term) || (lle_default && id != static_cast<int>(lib_loading_type::manual));
```
Switch terms for some optimization

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1074,6 +1063,19 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+
+				// Select default LLE sprx list for manual lib loading, otherwise all libs are unchecked and hidden
+				const bool lle_default = g_prx_list.count(text) && !g_prx_list.at(text);
+
```
Mismatch between comment and code...
The comment suggests:
const book selected = lle_default && id == static_cast<int>(lib_loading_type::manual);
items[i]->setCheckState(selected ? Qt::Checked : Qt::Unchecked);

elad335 replied:
Solved comments.

Megamouse replied:
It still doesn't make sense since you don't really uncheck them

elad335 commented:
* Fixed LLVM compilation of firmware. (regression)
* Removed lib loading types.

![image](https://user-images.githubusercontent.com/18193363/101980762-7a40c180-3c70-11eb-8101-44c4abc3e97d.png)

elad335 commented:
I also invalidated all old fw load settings, the default will be used after this version.

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1012,68 +1018,41 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
+		item->setCheckState(res && res != (lib.second * 2 - 1) ? Qt::Checked : Qt::Unchecked); // AND initialize check state
+
+		QString tip = lib.second ? 
+			QStringLiteral("Select to LLE. (HLE by default)") :
```
Why are these not tooltips in tooltips.h ?

Megamouse replied:
Or at least translated

elad335 replied:
Fixed.

Megamouse commented:
I haven't really looked at it yet. But is there a way to reset them?
Otherwise this is gonna get everyone in a world of pain

elad335 commented:
Empty list is default, manual reset is to unselect until empty. Theres no explicit way to reset it and there werent really any before.

elad335 commented:
I can add if requested of course.

elad335 commented:
Added a reset button.

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.ui:
```diff
@@ -2030,14 +1950,25 @@
-                  <horstretch>0</horstretch>
-                  <verstretch>0</verstretch>
-                 </sizepolicy>
+               <layout class="QHBoxLayout" name="horizontalLayout_2" stretch="1,0">
```
rename this layout

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -1012,68 +1018,36 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 		QListWidgetItem* item = new QListWidgetItem(qsv(lib.first), ui->lleList);
 		item->setFlags(item->flags() | Qt::ItemIsUserCheckable); // set checkable flag
-		item->setCheckState(Qt::Unchecked); // AND initialize check state
+		item->setCheckState(res && res != (lib.second * 2 - 1) ? Qt::Checked : Qt::Unchecked); // AND initialize check state
```
it's not really clear to the reader what you're doing here

elad335 replied:
Added comment.

elad335 commented on rpcs3/Emu/System.cpp:
```diff
@@ -1359,9 +1369,13 @@ game_boot_result Emulator::Load(const std::string& title_id, bool add_only, bool
 
 			const std::string game_path = "/dev_hdd0/game/" + m_path.substr(hdd0_game.size(), 9);
 
-			sys_log.notice("Forcing manual lib loading mode");
```
Someone smoked somthing real good when deciding HLE all libs for PS1 classics... well fixed.

Nekotekina replied:
I think it had its reasons at its time

elad335 commented:
One last addition before merge.
![image](https://user-images.githubusercontent.com/18193363/102010941-de838400-3d49-11eb-91ef-7263eb1c362a.png)
